docs/rules/sort-comp.md

```diff
@@ -9,7 +9,7 @@ When creating React components it is more convenient to always follow the same o
 The default configuration ensures that the following order must be followed:
 
   1. static methods and properties
-  2. lifecycle methods: `displayName`, `propTypes`, `contextTypes`, `childContextTypes`, `mixins`, `statics`,`defaultProps`, `constructor`, `getDefaultProps`, `getInitialState`, `state`, `getChildContext`, `componentWillMount`, `componentDidMount`, `componentWillReceiveProps`, `shouldComponentUpdate`, `componentWillUpdate`, `componentDidUpdate`, `componentWillUnmount` (in this order).
+  2. lifecycle methods: `displayName`, `propTypes`, `contextTypes`, `childContextTypes`, `mixins`, `statics`, `defaultProps`, `constructor`, `instance-variables`, `getDefaultProps`, `state`, `getInitialState`, `getChildContext`, `getDerivedStateFromProps`, `componentWillMount`, `UNSAFE_componentWillMount`, `componentDidMount`, `componentWillReceiveProps`, `UNSAFE_componentWillReceiveProps`, `shouldComponentUpdate`, `componentWillUpdate`, `UNSAFE_componentWillUpdate`, `getSnapshotBeforeUpdate`, `componentDidUpdate`, `componentWillUnmount`, `componentDidCatch` (in this order).
```

Comment:
I think `componentDidCatch` should be placed before `componentWillUnmount` since it is a method when component is "live"

Reply:
Where exactly would you recommend putting it? Do you think it would make more sense directly after `componentDidMount` or directly before `componentWillUnmount`?

I think part of the reason it was separated from the lifecycle and placed at the bottom in the docs is that it doesn't fall within the update cycle like the other methods between `componentDidMount` and `componentWillUnmount`, but I would agree that it would be strange to write it after `componentWillUnmount`.

Reply:
fwiw, you can throw in the `componentWillUnmount` of a child and it will be caught by `componentDidCatch`.

Reply:
I'm not sure, but you may even be able to throw an error it would catch from an unmounted child component, assuming you have an unhandled timeout somewhere. Putting `componentDidCatch` at the end does make it clear that it catches errors from the preceding lifecycle hooks, but then again it doesn't actually catch errors within the component it was defined inside of. I just went with the docs on this one originally since the existing ordering followed them, but I don't know whether they put it at the bottom to separate it documentation-wise or because they would actually suggest writing it at the bottom, since all of the examples I have seen have been error boundary components with no other lifecycle methods.

Reply:
It will be better to put `getDerivedStateFromProps` right before `static`, since it is a static lifecycle.

Reply:
`componentDidCatch` is a method with behavior very close to `componentDidUpdate`, I'd prefer it to be placed right after `componentDidUpdate`

Reply:
@yenshih That is not the ordering that the React documentation provides, as `getDerivedStateFromProps` is a lifecycle method, unlike other static methods and should be in a relevant position. Also `statics` only refers to the `React.createClass` statics object, and `static-methods` is actually where static methods on an ES6 class would be sorted to currently, which is before `statics` and actually lies at the very top of the component in the current default ordering.

@otakustay I don't really see the similarity there, `componentDidCatch` can be called at any time based on child component errors, whereas `componentDidUpdate` is called after every update to the component and is part of the update lifecycle. Also `componentDidCatch` is not something that is supposed to be triggered during the normal flow of your app, and it is explicitly stated in the documentation that it should not be used for control flow, which is what the lifecycle represents. I believe that that is the reason why React documentation has `componentDidCatch` at the bottom, after the normal lifecycle methods.

At this point I think it would be best to just stick with the React documentation's ordering since that seems to be what the ordering here has been based on in the past for the default configuration. If that ordering is not someone's preference that person can choose to override the default configuration to suit his or her needs.

Reply:
@joe-denea 

> getDerivedStateFromProps is invoked after a component is instantiated as well as when it receives new props

Since it is a replacement for `willMount` and `willReceiveProps`, it can be confused when being placed at the position of either `willMount` or `willReceiveProps`

I propose we order them like:

```
static displayName = '';

static defaultProps = {
};

... statics

static getDerivedStateFromProps() {
}

state = {
};

... instance fields
```

In this way we can easily have an overview about the structure of our `state` and how our state is synced with our `props`, the entire private state is clear before any life cycle occured

---

For `componentDidCatch`, I don't have a very strong insist, place it after `willUnmount` is OK

Reply:
Currently we may create initial state from props like:

```
constructor(props) {
  super(props);
  this.state = {
    foo: 1,
    bar: props.value
  };
}
```

Do we recommend, after `getDerivedStateFromProps` is available, to change it to:

```
state = {
  foo: 1,
  bar: null
};

static getDerivedStateFromProps(nextProps) {
  return {bar: nextProps.value};
}
```

In case we recommend "splitting self controlled and derived state", I'd prefer to place `getDerivedStateFromProps` after `constructor` so they have a better collabration

Reply:
Since the super constructor is run prior to instance variable initialization, you can reference `this.props` in the state definition and write what you have above as:

```JavaScript
state = {
  foo: 1,
  bar: this.props.value,
}
```

The code you wrote above with `getDerivedStateFromProps` is not equivalent to the code with `constructor` as it would update the state of `bar` when the `value` prop was updated.

`getDerivedStateFromProps` also receives the previous state as an argument, so it receives the initial state definition as an argument when the initial mount is going to be performed. It would therefore be backwards to the actual ordering of execution to place `getDerivedStateFromProps` prior to the initialization of `state`. That ordering would also be strange as defining instance variables and state as an instance variable is essentially shorthand for what is often done within a constructor in React, so you are proposing that `getDerivedStateFromProps`, a lifecycle method that is triggered on all component mounts and prop updates, be placed between an analogue for the constructor's functionality and the constructor itself.

Reply:
What I mean is that we should recommend the separation of "self controlled state" and those "derived from props" even they can be written in the form of constructor or class property, this may be a best practice:

```js
getDerivedStateFromProps(nextProps, prevState) {
  if (prevState.prevousPropValue === nextProps.value) {
    return {
      bar: value,
      // reset other states
    };
  }
}
```

In most case, we should not have different logics between initial and update state compute, one derived state property should always be synced to props

Reply:
My consideration is we should choose to place `getDerivedStateFromProps` either:

1. right after `constructor` if we think it should be responsible to compute the initial state
2. at the place where `componentWillReceiveProps` was if we recommend to use it only on prop changes (not initial instantiation)

Reply:
To be clear I think there is also a use case for the initial state declaration as a set of defaults, like the following

```JavaScript
state = {
  localeString: '0',
}

static getDerivedStateFromProps = (nextProps, prevState) => ({
  localeString: (!Math.isNaN(Number(nextProps.number)) && nextProps.number.toLocaleString())
    || prevState.localeString,
})
```

since this allows us to easily say that the component will convert the prop `number` to a locale string if it is a number and cache the value in state, and fall back to displaying the default of `'0'` or the last valid `number` that we received. In this case even if we wanted to avoid recalculating the value if the prop hadn't changed by caching the unmodified `number` prop in state for comparisons, we would have a sensible default of `0` for it since we wouldn't need to recalculate the default `localeString` if it were `0` initially.

If there is a sensible value for the initial state that isn't derived from props and the code reads clearly I would prefer to avoid setting the initial state based off of props in general, since it is an antipattern as it does not respect prop changes after mounting. I just would not leave out required state values from the initial state declaration as it serves to annotate the expected state fields in a component.

`getDerivedStateFromProps` remains a lifecycle method that is used for updates as well as mounting, and that is fired after the state initialization on the initial mount. If we were to place the state declaration below `getDerivedStateFromProps` in the code above it would appear as though `localeString` would come through as `undefined` in the initial `prevState` based on the ordering, which is not the case.

Reply:
Put didCatch directly before WillUnmount.